Added form fields for the admin podcasts dashboard.

wotta opened the pull request:
<!--
     For Work In Progress Pull Requests, please use the Draft PR feature,
     see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.

     For a timely review/response, please avoid force-pushing additional
     commits if your PR already received reviews or comments.

     Before submitting a Pull Request, please ensure you've done the following:
     - 📖 Read the Forem Contributing Guide: https://github.com/forem/forem/blob/master/CONTRIBUTING.md#create-a-pull-request.
     - 📖 Read the Forem Code of Conduct: https://github.com/forem/forem/blob/master/CODE_OF_CONDUCT.md.
     - 👷‍♀️ Create small PRs. In most cases this will be possible.
     - ✅ Provide tests for your changes.
     - 📝 Use descriptive commit messages.
     - 📗 Update any related documentation and include any relevant screenshots.
-->

## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
Adding more edit capabilities in the admin dashboard for podcasts.

## Related Tickets & Documents
Closes https://github.com/forem/forem/issues/10482

## QA Instructions, Screenshots, Recordings
When this issue is being pulled for a functional review one should have access to an admin 

## Added tests?

- [ ] yes
- [ ] no, because they aren't needed
- [x] no, because I need help _at least I think_

## Added to documentation?

- [ ] docs.forem.com
- [ ] readme
- [x] no documentation needed

## [optional] Are there any post deployment tasks we need to perform?
Not that I know of 

## [optional] What gif best describes this PR or how it makes you feel?

![hoping this is what they wanted :D](https://media.giphy.com/media/xULW8hXZZVcxS2bdDi/giphy.gif)

fdoxyz commented:
Thanks for submitting the PR @wotta. The fields are there but the update doesn't seem to work yet.

From what I can tell you'll still need to add the new params into the [list of permitted params](https://github.com/fdoxyz/forem/blob/master/app/controllers/admin/podcasts_controller.rb#L66). It would also be good to [update the test](https://github.com/fdoxyz/forem/blob/master/spec/requests/admin/podcasts_spec.rb#L62-L76) that validates this functionality

rhymes commented on app/views/admin/podcasts/edit.html.erb:
```diff
@@ -64,6 +64,62 @@
+        <%= f.label :image, for: "image" %>
+        <%= f.file_field :image, accept: "image/*", class: "form-control" %>
+      </div>
+      <div class="crayons-field crayons-field--checkbox">
+        <%= f.check_box :reachable, class: "crayons-checkbox" %>
+        <%= f.label :reachable, "Podcast is reachable", class: "crayons-field__label" %>
+      </div>
```
This makes me wonder if `podcast.reachable` is even used in the code or we just use the scope `Podcast.episode` which in turn uses `podcast_episode.reachable`

cc @lightalloy

rhymes reviewed (commented):
Thanks @wotta!

wotta commented:
@fdoxyz From what I saw on the "migration" not all columns are required.
Would I still need to add all the form fields to `podcast` params?

wotta commented:
Could somebody help me out with tests?
I added the parameters but it expects some whole different information than I added.

It doesn't seem logical at this moment when I look at it so I guess I am missing something.
@rhymes @fdoxyz

fdoxyz commented on spec/requests/admin/podcasts_spec.rb:
```diff
@@ -62,15 +62,56 @@
   describe "Updating" do
```
A reusable variable for the params to pass in to each test case

```suggestion
  describe "Updating" do
    let(:update_params) do
      {
        title: "hello",
        feed_url: "https://pod.example.com/rss.rss",
        description: "Description",
        itunes_url: "https://itunes.example.com",
        overcast_url: "https://overcast.example.com",
        android_url: "https://android.example.com",
        soundcloud_url: "https://soundcloud.example.com",
        website_url: "https://example.com",
        twitter_username: "@ThePracticalDev",
        pattern_image: fixture_file_upload("files/800x600.png", "image/png"),
        main_color_hex: "ffffff",
        image: fixture_file_upload("files/podcast.png", "image/png")
        slug: "postcast-test-url",
        reachable: true,
        published: true,
      }
    end

```

fdoxyz commented on spec/requests/admin/podcasts_spec.rb:
```diff
@@ -62,15 +62,56 @@
   describe "Updating" do
     it "updates" do
       put admin_podcast_path(podcast), params: { podcast: { title: "hello",
-                                                            feed_url: "https://pod.example.com/rss.rss" } }
+                                                            feed_url: "https://pod.example.com/rss.rss",
+                                                            description: "Description",
+                                                            itunes_url: "https://itunes.example.com",
+                                                            overcast_url: "https://overcast.example.com",
+                                                            android_url: "https://android.example.com",
+                                                            soundcloud_url: "https://soundcloud.example.com",
+                                                            website_url: "https://example.com",
+                                                            twitter_username: "@ThePracticalDev",
+                                                            pattern_image: "https://dummyimage.com/400x400.jpg",
+                                                            main_color_hex: "#ffffff",
+                                                            slug: "postcast-test-url",
+                                                            image: "https://dummyimage.com/400x400.jpg",
+                                                            reachable: true,
+                                                            published: true
+                                                          } }
```
Now the update can be replaced with something like this:

```ruby
put admin_podcast_path(podcast), params: { podcast: update_params }
```

fdoxyz commented on spec/requests/admin/podcasts_spec.rb:
```diff
@@ -62,15 +62,56 @@
+      expect(podcast.soundcloud_url).to eq("https://soundcloud.example.com")
+      expect(podcast.website_url).to eq("https://example.com")
+      expect(podcast.twitter_username).to eq("@ThePracticalDev")
+      expect(podcast.pattern_image).to eq("https://dummyimage.com/400x400.jpg")
+      expect(podcast.main_color_hex).to eq("#ffffff")
+      expect(podcast.slug).to eq("postcast-test-url")
+      expect(podcast.image).to eq("https://dummyimage.com/400x400.jpg")
```
Since a separate test would validate the images being updated there's no need to check for them here. Also important to update the expected value for `main_color_hex`

```suggestion
      expect(podcast.main_color_hex).to eq("ffffff")
      expect(podcast.slug).to eq("postcast-test-url")
```

fdoxyz commented on spec/requests/admin/podcasts_spec.rb:
```diff
@@ -62,15 +62,56 @@
 
     it "redirects after update" do
       put admin_podcast_path(podcast), params: { podcast: { title: "hello",
-                                                            feed_url: "https://pod.example.com/rss.rss" } }
+                                                            feed_url: "https://pod.example.com/rss.rss",
+                                                            description: "Description",
+                                                            itunes_url: "https://itunes.example.com",
+                                                            overcast_url: "https://overcast.example.com",
+                                                            android_url: "https://android.example.com",
+                                                            soundcloud_url: "https://soundcloud.example.com",
+                                                            website_url: "https://example.com",
+                                                            twitter_username: "@ThePracticalDev",
+                                                            pattern_image: "https://dummyimage.com/400x400.jpg",
+                                                            main_color_hex: "#ffffff",
+                                                            slug: "/test",
+                                                            image: "https://dummyimage.com/400x400.jpg",
+                                                            reachable: true,
+                                                            published: true
+                                                          } }
```
Again we could now reuse the `update_params`

```ruby
put admin_podcast_path(podcast), params: { podcast: update_params }
```

fdoxyz reviewed (changes requested):
Hey @wotta absolutely! Sorry for the delay in following up, I'm just returning from a few days offline. There's a few things I see that are happening in the tests:

1. The `main_color_hex` value for Podcasts is validated against a Regex and the value shouldn't contain the hash char
1. `/test` is an invalid slug (for the `"redirects after update"` spec)
1. Both `pattern_image` & `image` need to be sent as if they were just uploaded using a file (passing the resulting URL won't work in this case)

I think 1 & 2 are easy to fix with the current specs we have in place, but testing the images may be easier to do if we add another test for them separately. If anyone can suggest a better way to do this it would be great, but I'm leaving my suggestions inline.

This would be the new test to include inside the same `describe "Updating" do` in order to test that images were updated (since I'm suggesting to remove the check from the existing test)

```ruby
    it "updates image & pattern_image" do
      expect do
        put admin_podcast_path(podcast), params: { podcast: update_params }
        podcast.reload
      end.to change(podcast, :image) && change(podcast, :pattern_image)
    end
```

Let me know if you need me to further clarify anything else 🙂

wotta commented:
Thanks for the tips! I will try to update PR tonight when.

wotta commented:
Hello everyone,

I see the following error. 
Now I don't mind fixing the error if needed but I am not sure if I caused the error.

```
Failures:
  1) Creating an article with the editor creates a new article
     Failure/Error: expect(page).to have_selector("header h1", text: "Sample Article")
       expected to find css "header h1" but there were no matches
```

fdoxyz commented:
The error wasn't related to the changes. I restarted the build and now it passed. I'm hoping to review this later today.

fdoxyz commented on spec/requests/admin/podcasts_spec.rb:
```diff
@@ -60,17 +60,66 @@
+      }
+    end
+
+    let(:update_params) do
```
This `update_params` is repeated, we only need one.

fdoxyz commented on spec/requests/admin/podcasts_spec.rb:
```diff
@@ -60,17 +60,66 @@
+      expect(podcast.reachable).to eq(true)
+      expect(podcast.published).to eq(true)
     end
 
```
We can add this test right in between these two to validate the image test:


```ruby
it "updates image & pattern_image" do
  expect do
    put admin_podcast_path(podcast), params: { podcast: update_params }
    podcast.reload
  end.to change(podcast, :image) && change(podcast, :pattern_image)
end
```

fdoxyz reviewed (changes requested):
Thanks for the changes @wotta, almost there! I left two small comments inline

fdoxyz commented:
Hi @wotta, just wanted to let you know a small changes was added in the last week and you'll need to fix the conflicts. Nothing major, I think you're safe keeping your changes on all of them. Let me know if you need any help with this 🙂

wotta commented:
Hello @fdoxyz,

It has been too hectic to even respond because of a lot of stuff happening.
I can try to fix everything tonight.

And hopefully, this issue can be resolved finally.

wotta commented:
Good morning @fdoxyz,

I updated the PR with the master branch but I am getting the following errors:
![Screen Shot 2020-11-02 at 8 01 45 AM](https://user-images.githubusercontent.com/13332240/97839095-c992ea00-1ce1-11eb-83ed-d3b53bc114aa.png)

Is this something that needs to be fixed by me or should a different PR be made to fix those errors?

rhymes commented:
@wotta I merged master in, as it contains the fix for that error

wotta commented:
Aah thank you. When I merged the master branch it didn't have that fix.

Also wasn't sure how to act on this because of the scale of this project.

Thank you.

fdoxyz reviewed (approved):
Thank you @wotta for making the changes! It LGTM now 👍

wotta commented:
Thank you @fdoxyz!

Who is supposed to merge it at this point?

fdoxyz commented:
Our process requires 2 approvals in order to merge. As soon as another team member weighs in and if everything looks good we'll be ready to merge :)

rhymes reviewed (approved):
Thank you @wotta
